Fix some crashes seen with site isolation

achristensen07 opened the pull request:
#### 6028f9871d7caffc878f2dbafcc692dc85be3ed6
<pre>
Fix some crashes seen with site isolation
<a href="https://bugs.webkit.org/show_bug.cgi?id=286866">https://bugs.webkit.org/show_bug.cgi?id=286866</a>

Reviewed by Simon Fraser.

This fixes two crashes I&apos;ve seen when running tests with site isolation on.
The first is in EventHandler::scrollRecursively if the parent frame is a RemoteFrame.
The second is in WebPageProxy::didCreateSubframe if the parent frame has
been destroyed but the process from which we received this message has not
yet received that update before sending the message to destroy the child
frame.  This is a new race condition with site isolation having multiple
web content processes, and it is correct to do nothing in this case, which
can happen during teardown sometimes.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::scrollRecursively):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCreateSubframe):

Canonical link: <a href="https://commits.webkit.org/289687@main">https://commits.webkit.org/289687@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/544c68c7ead2ea142f61f02beafe32bf1995b9be

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87698 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7214 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42082 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92563 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38444 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89749 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7595 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15385 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67708 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25454 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90700 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5793 "Found 1 new test failure: fast/forms/ios/focus-input-in-fixed.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79352 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48080 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5580 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33754 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37555 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75978 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34636 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94449 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14866 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10906 "Found 1 new test failure: ipc/create-media-source-with-invalid-constraints-crash.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76559 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15120 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75208 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75794 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20158 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18592 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7826 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13675 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14882 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20184 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14626 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18070 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16408 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -1441,6 +1441,8 @@ bool EventHandler::scrollRecursively(ScrollDirection direction, ScrollGranularit
     if (!parent)
         return false;
     RefPtr localParent = dynamicDowncast<LocalFrame>(parent.get());
+    if (!localParent)
```
This one should be easy to test, with a `scrollIntoView()` in a remote subframe.

achristensen07 replied:
That doesn't seem to hit it.  Not sure why.

achristensen07 replied:
This is only called from executeScrollLineUp, executeScrollLineDown, and WebPage::scroll

achristensen07 committed:
Fix some crashes seen with site isolation
https://bugs.webkit.org/show_bug.cgi?id=286866

Reviewed by Simon Fraser.

This fixes two crashes I've seen when running tests with site isolation on.
The first is in EventHandler::scrollRecursively if the parent frame is a RemoteFrame.
The second is in WebPageProxy::didCreateSubframe if the parent frame has
been destroyed but the process from which we received this message has not
yet received that update before sending the message to destroy the child
frame.  This is a new race condition with site isolation having multiple
web content processes, and it is correct to do nothing in this case, which
can happen during teardown sometimes.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::scrollRecursively):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCreateSubframe):

Canonical link: https://commits.webkit.org/289687@main
